### PR TITLE
feat(sql): add variant of timestamp_floor supporting an offset origin

### PR DESCRIPTION
**Pull request by nwoolmer**
This is a pre-requisite for https://github.com/questdb/questdb/pull/4733

`timestamp_floor(sN)` calculates the intervals based on a single origin, `0` i.e `1970-01-01`. This PR adds support for a third parameter, `offset`, which acts as a new origin.

The signature is: `timestamp_floor(sNn)`.

This rationalises the function's behaviour when being used to calculate intervals/bins.

Before:

```sql
select timestamp_floor('5d', '2018-01-01T00:00:00.000000Z')
```

`2017-12-30T00:00:00.000000Z`

After:

```sql
select timestamp_floor('5d', '2018-01-01T00:00:00.000000Z', '2018-01-01T00:00:00.000000Z')
```

`2018-01-01T00:00:00.000000Z`

Also renamed `addYear` to `addYears` so its consistent with similar functions.

Data example:

Before:

```sql
select timestamp_floor('5d', pickup_datetime) t, count from trips where pickup_datetime in '2018'
order by 1;
```
image

After:

```sql
select timestamp_floor('5d', pickup_datetime, '2018') t, count from trips where pickup_datetime in '2018'
order by 1;
```

image

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 121 / 126 (96.03%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|33|36|91.67%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctions.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 141 / 146 (96.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|51|54|94.44%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctions.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 141 / 146 (96.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|51|54|94.44%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctions.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 151 / 177 (85.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|12|08.33%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|9|19|47.37%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|51|54|94.44%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctions.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 141 / 146 (96.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|51|54|94.44%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctions.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

File: core/src/main/java/io/questdb/std/datetime/microtime/Timestamps.java [hunk 1]
```diff
@@ -1103,8 +1216,12 @@ private static long getTimeMicros(long micros) {
     }
 
     private static long getTimeMicros(long micros, int stride) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/datetime/microtime/Timestamps.java:1218]: nit: unused function

**Comment by jerrinot** [comment]
@nwoolmer Did you consider benchmarking it? I’d assume functions in `Timestamps` can often be on a hot-path (flooring timestamps for every row). e.g. `getTimeMicros()`

The code is now more complex (bytecode size impacts inlining decisions and inlining impacts on other JIT optimizations) There also is an extra subtraction.

I can see TimestampFloorXYFunction often do this anyway:
```java
public long floor(long timestamp) {
    return Timestamps.floorDD(timestamp, stride);
    return offset == 0 ? Timestamps.floorDD(timestamp, stride) : Timestamps.floorDD(timestamp, stride, offset);
}
```
So there is a fork and later it's merged when `Timestamps.floorDD(timestamp, stride)` delegates to `Timestamps.floorDD(timestamp, stride, 0)` anyway. 

So if you already have the branch then you could have specialized flooring functions in Timestamps. But this is hard to tell w/o real-world-like benchmarks. WDYT?

**Comment by jerrinot** [comment]
@nwoolmer how would you define semantic of `Timestamp.floorXY()` when `offset > micros` ?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 141 / 146 (96.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|51|54|94.44%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctions.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

**Review by jerrinot — DISMISSED** [review]
looking good, it can be merged after:
1. the negative micros are explored/clarified as discussed. the code is currently sending mixed signals.
2. consider a benchmark to make sure there is no regression due to more complex code on a hot path.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 106 / 130 (81.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctions.java|41|55|74.55%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|36|44|81.82%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorOffsetFunctionFactory.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampAddFunctionFactory.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**